fix(source): change kafka "offset" to be left inclusive and right exclusive

stdrc opened the pull request:
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

https://www.notion.so/risingwave-labs/Which-s-Next-The-Source-Offset-Confusion-17ac5498bc3640a383b21a7ea8bbee58?pvs=4

Previously what we stored to `start_offset` field in `KafkaSplit` metadata is actually "last seen/consumed offset", and this had two problems:

1. When `scan.startup.timestamp.millis` is set, we have to subtract 1 from the offset derived from the given timestamp.
2. When initializing a Kafka reader, we have to add 1 back to the offset.

When these are mixed with Kafka low/high watermark, the code became very hard to understand and correctly maintain.

This PR gives the `start_offset` and `stop_offset` concepts a clear definition, i.e. when start to consume from Kafka, `start_offset` represents the first offset to read (inclusive), `stop_offset` represents the last offset to read PLUS 1 (exclusive). The new definitions match the low/high watermark definition in Kafka, so hopefully can be more intuitive.

To include some background of other `XxxSplit` structs (https://github.com/risingwavelabs/risingwave/pull/16256#discussion_r1560705754):

> IMO we should `+1` if possible when `update_offset` (unfortunately some source platforms don't allow simply doing addition on offset), that's in my mind the most elegant and clear solution. But in our current actual implementation, we have 3 cases:
>
> 1. `+1` when creating reader (Kafka), will be fixed in https://github.com/risingwavelabs/risingwave/pull/16257
> 2. Set offset to `AfterSequenceNumber(last_seen_offset)` (Kinesis), correct and exactly once.
> 3. Read from `last_seen_offset` on recovery, meaning duplicated items (Debezium CDC). The duplication seem unavoidable, see https://debezium.io/documentation/reference/stable/development/engine.html#_handling_failures.

## Checklist

- [x] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)
<!-- To manually trigger a benchmark, please check out [Notion](https://www.notion.so/risingwave-labs/Manually-trigger-nexmark-performance-dashboard-test-b784f1eae1cf48889b2645d020b6b7d3). -->
- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.


<!--
Please create a release note for your changes.

Discuss technical details in the "What's changed" section, and
focus on the impact on users in the release note.

You should also mention the environment or conditions where the impact may occur.
-->

</details>

stdrc commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/16257?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#16257** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/16257?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#16256** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/16256?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @stdrc and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

stdrc commented on src/connector/src/source/kafka/split.rs:
```diff
@@ -33,35 +69,62 @@ impl SplitMetaData for KafkaSplit {
     fn update_offset(&mut self, last_seen_offset: String) -> ConnectorResult<()> {
-        self.start_offset = Some(last_seen_offset.as_str().parse::<i64>().unwrap());
+        let last_seen_offset = last_seen_offset.as_str().parse::<i64>().unwrap();
+        self.start_offset = Some(last_seen_offset + 1);
```
After the fixing, we only need to add 1 here.

stdrc commented:
> but I think it is not worthy introducing the complexity distinguishing the behavior because the kafka split reader has limited behaviors, reading from beginning, from a timestamp, and from latest offset. No more expected behaviors or extensible usage (from rw side and kafka side). Making the code happen to work correctly is enough.
> I admit that we did not cover all behaviors well in prev impl and left some bugs. But the actual case is no user finds them. So I wonder just fixing the bugs is enough and there is no need for a new schema. What do you think?

Don't you feel uncomfortable when seeing so many confusing `-1` and `+1` in the code?

<img width="1130" alt="截屏2024-04-11 17 26 09" src="https://github.com/risingwavelabs/risingwave/assets/5317095/4458bef6-4fe7-4574-872f-47590a5d31a6">
<img width="1127" alt="截屏2024-04-11 17 26 05" src="https://github.com/risingwavelabs/risingwave/assets/5317095/a95a297d-5bd8-4a61-bd08-6e0774f3798c">
<img width="1129" alt="截屏2024-04-11 17 25 44" src="https://github.com/risingwavelabs/risingwave/assets/5317095/b8e7ab56-e80c-4f0b-9a7c-753928a00361">
<img width="1123" alt="截屏2024-04-11 17 25 38" src="https://github.com/risingwavelabs/risingwave/assets/5317095/996a6c11-6b7b-4fb6-9eb0-661bd3c8f6d8">
<img width="1115" alt="截屏2024-04-11 17 25 33" src="https://github.com/risingwavelabs/risingwave/assets/5317095/6c1984c1-6e95-449e-b80a-c055798a5288">
<img width="1126" alt="截屏2024-04-11 17 25 27" src="https://github.com/risingwavelabs/risingwave/assets/5317095/0c6a0b67-6b59-4946-987c-eaa4ef21c667">

stdrc commented:
Also, the changes to `KafkaSplit` in this PR are actually making `KafkaSplit` more consistent with other `XxxSplit`. That is, making the `start_offset` real **start offset**. You can check `KinesisSplit` and `CdcSplitBase`, you will find that the `start_offset` in them is really the **start offset** when starting to read from the source on recovery. For the latter Debezium CDC case, it reads duplicated item and that is documented.

tabVersion commented:
> Also, the changes to `KafkaSplit` in this PR are actually making `KafkaSplit` more consistent with other `XxxSplit`. That is, making the `start_offset` real **start offset**. You can check `KinesisSplit` and `CdcSplitBase`, you will find that the `start_offset` in them is really the **start offset** when starting to read from the source on recovery. For the latter Debezium CDC case, it reads duplicated item and that is documented in Debezium documents.

I understand the improvement and I am up with the overall goal of making rw more maintainable.

My point is that it changes the schema in the state table, no matter whether simple or not, it is a **risk** and I don't see enough benefits for taking it. 
For "maintainability", as you can see in the `kafka source enumerator` history for the past year, there is no actual logic change. This part is not evolving. This part has done all it needs to do, so I think the ideal way is to fix the bug rather than bring new abstracts and take the compatibility issue. 
If there was a chance to design `source_v2`, I would be happy to bring the proposal to production.

stdrc commented:
> My point is that it changes the schema in the state table, no matter whether simple or not, it is a **risk** and I don't see enough benefits for taking it.

Does that justify the bug fixed by #16073 which had existed for 2 years, from the 1st day when Kafka source was introduced (https://github.com/risingwavelabs/risingwave/commit/7678f25a769967874ba5802e00027f4a8ca21cd6)? 

Or the bug fixed by #6647, in which you found reading from "latest" was not working as expected because we forgot to subtract 1 from the high watermark to construct the "start offset"?

Or #6725 that introduced batch query on Kafka source, in which the following code snippet actually broke the definition of "watermark" in Kafka?

```rs
        let mut watermarks = {
            let mut ret = HashMap::new();
            for partition in &topic_partitions {
                let (low, high) = self
                    .client
                    .fetch_watermarks(self.topic.as_str(), *partition, KAFKA_SYNC_CALL_TIMEOUT)
                    .await?;
                ret.insert(partition, (low - 1, high));
            }
            ret
        };
```

The reason of all these bugs and semantic mismatch is because of the wrongly defined `start_offset`. As you said, things just barely **happened to work**. But IMO we need things to **confidently work**. We need to understand each piece of our code. Although now it seem peaceful, that may be just because some combination of the functionalities are not used by any user (or they just didn't report to us?). Instead of regarding the changes in this PR a "risk", **I regard the semantic mismatch a risk**, which can potentially cause new bugs or reveal old ones every time we extend or just refactor the module.

Back to this PR, it doesn't introduce any backward-incompatibility. It seamlessly migrate any old `KafkaSplit` value to fit the new code. Especially for following change pair (migrate existing offset by adding 1, remove `+1` when create reader) in happy path, I am personally certain that the changes are correct.

```diff
--- a/src/connector/src/source/kafka/source/reader.rs
+++ b/src/connector/src/source/kafka/source/reader.rs
@@ -110,7 +110,7 @@ impl SplitReader for KafkaSplitReader {
                 tpl.add_partition_offset(
                     split.topic.as_str(),
                     split.partition,
-                    Offset::Offset(offset + 1),
+                    Offset::Offset(offset),
                 )?;
             } else {
                 tpl.add_partition(split.topic.as_str(), split.partition);

```

```diff
--- a/src/connector/src/source/kafka/split.rs
+++ b/src/connector/src/source/kafka/split.rs
@@ -22,8 +22,44 @@ use crate::source::{SplitId, SplitMetaData};
 pub struct KafkaSplit {
     pub(crate) topic: String,
     pub(crate) partition: i32,
+
+    /// The offset pointing to the first message to read from the partition. Inclusive.
     pub(crate) start_offset: Option<i64>,
+    /// The offset pointing to the next position of the last message to read from the partition. Exclusive.
     pub(crate) stop_offset: Option<i64>,
+
+    /// The version of the split metadata. `None` means version 0.
+    version: Option<u32>,
+}
+
+/// The current version of the split metadata.
+/// Please increment this version and write a migration script below when you make a change to
+/// the split metadata.
+const CURRENT_VERSION: u32 = 1;
+
+static MIGRATION_SCRIPTS: &[fn(serde_json::Value) -> serde_json::Value] =
+    &[/* 0: 0 -> 1 */ migration::v0_to_v1];
+
+mod migration {
+    pub fn v0_to_v1(mut value: serde_json::Value) -> serde_json::Value {
+        // In v0, the `start_offset` field represented the last seen offset, which means when
+        // initializing a reader from the split, we should add 1 to the offset. But in the case
+        // of timestamp offset, it happened that we forgot to subtract 1 from the offset
+        // when creating `KafkaSplit`, causing the first message to be missed.
+        // See https://github.com/risingwavelabs/risingwave/issues/16046 for more.
+        //
+        // In v1, we normalize the `start_offset` field to represent the first message to read
+        // from the partition. So we need to add 1 to any v0 `start_offset` field. We also
+        // normalize the `stop_offset` field to represent the next position of the last message,
+        // while fortunately, although not documented, the v0 `stop_offset` field was already
+        // representing the next position of the last message.
+
+        let start_offset = value["start_offset"]
+            .as_i64()
+            .map(|start_offset| start_offset + 1);
+        value["start_offset"] = start_offset.into();
+        value
+    }
 }
```

BugenZhao commented on src/connector/src/source/kafka/split.rs:
```diff
@@ -22,8 +22,44 @@ use crate::source::{SplitId, SplitMetaData};
+static MIGRATION_SCRIPTS: &[fn(serde_json::Value) -> serde_json::Value] =
+    &[/* 0: 0 -> 1 */ migration::v0_to_v1];
+
+mod migration {
```
I would proposing another compatibility approach: 😄

```rust
#[derive(serde::Serialize, serde::Deserialize, Debug)]
#[serde(untagged)]
enum StartOffset {
    V0 {
        #[serde(rename = "start_offset")]
        last_seen_offset: i64,
    },
    V1 {
        #[serde(rename = "start_offset_v1")]
        start_offset: i64,
    },
}

#[derive(serde::Serialize, serde::Deserialize, Debug)]
struct KafkaSplit {
    topic: String,
    partition: i32,

    /// The offset pointing to the first message to read from the partition. Inclusive.
    #[serde(flatten)]
    start_offset: Option<StartOffset>,
    /// The offset pointing to the next position of the last message to read from the partition. Exclusive.
    stop_offset: Option<i64>,
}
```

tabVersion commented:
@stdrc and I had a thorough offline discussion, recognizing the core issue stems from the differing semantics of the start offset during source creation and recovery.

This discrepancy has led to frequent +1/-1 adjustments across Kafka implementations. While this PR aims to standardize the definition of start offset across various connectors, it appears to users as altering the state table's definition.
Currently, the state table's start offset represents the offset of the last confirmed message, but with this change, it would represent the offset for the next message to be fetched, effectively shifting to offset +1 and other connectors stay the same. This creates an inconsistency for users, urging us to find a better solution.

fuyufjh commented:
In fact, I am more opposed to the previous PR (renaming `last_offset -> last_seen_offset`). Commented there: https://github.com/risingwavelabs/risingwave/pull/16256#discussion_r1562234832

StrikeW commented on src/connector/src/source/kafka/enumerator/client.rs:
```diff
@@ -310,10 +314,10 @@ impl KafkaSplitEnumerator {
                 Offset::Offset(offset) => {
-                    // XXX(rc): currently in RW source, `offset` means the last consumed offset, so we need to subtract 1
-                    result.insert(elem.partition(), Some(offset - 1));
+                    result.insert(elem.partition(), Some(offset));
```
Here why we need to `-1` before?
The doc of [rd_kafka_offsets_for_times](https://docs.confluent.io/platform/7.5/clients/librdkafka/html/rdkafka_8h.html#ae308aaf534807cfd4c71e423fc214929) said:
```
The returned offset for each partition is the earliest offset whose timestamp is greater than or equal to the given timestamp in the corresponding partition.
```
So I think we just need to start the consumption from the returned offset, not need to `-1`.

stdrc replied:
Haha, your question is exactly the reason why I created this PR. See? No one can understand these -1/+1 without explanation or deep dive into the code. This is a maintenance hell! See https://github.com/risingwavelabs/risingwave/pull/16257#issuecomment-2060271006 for more explanation.

StrikeW replied:
I have read your [explanation](https://github.com/risingwavelabs/risingwave/pull/16257#issuecomment-2060271006) below. IIUC, we `-1` here to create a `Split`, then when the `Split` passed to source executor, it will `+1` to create a `Reader` start consumption, right?

stdrc commented:
Let me clarify the problem with a more concise language.

For `KafkaSplit` and also some other such `XxxSplit`s, the `start_offset` field has two different semantics in the following two situations:

1. When a split struct is created by `Enumerator`, the `start_offset` means **the offset of the first message we should read**.
2. When `update_offset` is called (by `SourceExecutor`), the `start_offset` is **replaced with a “last seen offset”** which is got from the stream chunk returned by `Reader`.

This semantic mismatch is the key problem.

Now our Kafka source does work, because we made the semantics in the 1st situation align with the 2nd one, **by subtracting 1 from the offset when creating a `Split` and adding 1 to the offset when creating/recovering a `Reader`**. However the solution is very counter-intuitive. By its name, the `start_offset` should’ve store the “start offset” or “next offset”, but it actually stores “last seen offset” or “previous offset”. And that also caused other source implementations to follow the error, because new sources just copied the design from old ones, especially file source, which is acting totally wrong.

I think a more reasonable and intuitive solution should be align the 2nd semantics with the 1st one, which is exactly the way how Flink, Kafka and plenty of other industrial products handle offsets. **That is, to be more clear, always storing the “next offset” we should read next time.** And this is also the **only** way that can correctly work for file source.

xxchan commented:
Let's reopen when the work is resumed 🤫

stdrc commented:
> Let's reopen when the work is resumed 🤫

Sure😁
